releng_sop/koji_sign.py

```diff
@@ -0,0 +1,715 @@
+        :type  rpm_info_list: list
+        :return: {rpm_id: {sigkey: sighash}}
+        :rtype:  dict
+        """
+        result = {}
+
+        self.koji_session.multicall = True
+        for rpm_info in rpm_info_list:
+            self.koji_session.queryRPMSigs(rpm_info["id"])
+        data = self.koji_session.multiCall()
+
+        for [sig_list] in data:
+            for sig in sig_list:
+                rpm_id = sig["rpm_id"]
+                sigkey = sig["sigkey"].lower()
+                sighash = sig["sighash"]
+                result.setdefault(rpm_id, {})[sigkey] = sighash
+        return result
+
+    def find_cached(self, rpm_info_list, rpm_sig_dict, sigkeys):
+        """
+        Split rpm_info_list into (cached, uncached) according to cached sigs found in koji.
+
+        :param rpm_info_list: List of koji rpm_info dictionaries
+        :type  rpm_info_list: list
+        :param rpm_sig_dict: A dictionary obtained from get_rpm_sig_dict() method
+        :type  rpm_sig_dict: dict
+        :param sigkeys: List of sigkeys
+        :type  sigkeys: list
+        :return: (cached, uncached) with rpm_info lists
+        :rtype:  tuple
+        """
+        cached = []
+        uncached = []
+
+        sigkeys = set([i.lower() for i in sigkeys])
+        for rpm_info in rpm_info_list:
+            existing_sigs = set(rpm_sig_dict.get(rpm_info["id"], []))
+            if existing_sigs & sigkeys:
+                cached.append(rpm_info)
+            else:
+                uncached.append(rpm_info)
+
+        return cached, uncached
+
+    def _get_rpm_path(self, rpm_info, sigkey):
+        """
+        Return path to a signed or unsigned copy of an RPM in koji.
+
+        :param rpm_info: Koji rpm_info dictionary
+        :type  rpm_info: dict
+        :param sigkey: Sigkey. Empty string or None for unsigned.
+        :type  sigkey: str
+        :return: Path to an RPM
+        :rtype:  str
+        """
+        if sigkey:
+            # signed
+            path = os.path.join(self.koji_module.pathinfo.build(rpm_info["build"]), self.koji_module.pathinfo.signed(rpm_info, sigkey))
+        else:
+            # unsigned
+            path = os.path.join(self.koji_module.pathinfo.build(rpm_info["build"]), self.koji_module.pathinfo.rpm(rpm_info))
+        return path
+
+    def _find_rpms(self, rpm_info_list, func):
+        """
+        A generic method for RPM lookups.
+
+        :param rpm_info_list: List of koji rpm_info dictionaries
+        :type  rpm_info_list: list
+        :param func: A match function that takes 2 arguments: (rpm_info, matches_by_rpm_id)
+        :type  func: function
+        :return: (matched, unmatched) koji rpm_info dictionaries
+        :rtype:  tuple
+        """
+        matched = []
+        unmatched = []
+        matches_by_rpm_id = {}
+
+        def _wrapped_func(rpm_info):
+            return func(rpm_info, matches_by_rpm_id)
+
+        pool = multiprocessing.dummy.Pool(10)
```

Comment:
This line suggests that multiprocessing is used, although dummy stands for threading implementation. It would worth to mention this in a comment to avoid confusion about missing IPC method.